File: Job Sequencing Problem.cpp
class Solution 
{
    public:
    vector<int> JobScheduling(Job arr[], int n) 
    { 
       vector<int>ans;
       sort(arr,arr+n,[&](Job a,Job b){
           return (a.profit == b.profit)?a.dead < b.dead:a.profit > b.profit;
       });
       int ct = 0;
       int profit = 0;
       vector<int>timeline(n+1,0);
       for(int i=0;i<n;i++){
           int st = arr[i].dead-1;
           if(st < 0)continue;
           while(timeline[st]==1){
               st--;
           }
           if(st < 0)continue;
           timeline[st] = 1;
           ct++;
           profit += arr[i].profit;
       }
       ans.push_back(ct);
       ans.push_back(profit);
       return ans;
    } 
};

File: Job Sequencing Problem.py
class Solution:
    def jobSequencing(self, deadline, profit):
        # code here
        from heapq import heappush, heappop
        jobs = list(zip(deadline, profit))
        jobs.sort()
        max_profit = 0
        job_count = 0
        max_heap = []
        index = len(jobs) - 1
        
        for i in range(jobs[-1][0], 0, -1):
            while index >= 0 and jobs[index][0] == i:
                heappush(max_heap, -jobs[index][1]) 
                index -= 1
            
            if max_heap:
                job_count += 1
                max_profit += -heappop(max_heap)
        
        return [job_count, max_profit]
 
